runtime/sema/check_composite_declaration.go

```diff
@@ -1963,6 +1996,99 @@ func (checker *Checker) enumMembersAndOrigins(
 	return
 }
 
+func (checker *Checker) checkDefaultDestroyEvent(
+	eventType *CompositeType,
+	eventDeclaration ast.CompositeLikeDeclaration,
+	containerType ContainerType,
+	containerDeclaration ast.Declaration,
+) {
+
+	// an event definition always has one "constructor" function in its declaration list
+	members := eventDeclaration.DeclarationMembers()
+	functions := members.SpecialFunctions()
+	constructorFunctionParameters := functions[0].FunctionDeclaration.ParameterList.Parameters
+
+	var checkParamExpressionKind func(ast.Expression)
+	checkParamExpressionKind = func(arg ast.Expression) {
+		switch arg := arg.(type) {
+		case *ast.StringExpression,
+			*ast.BoolExpression,
+			*ast.NilExpression,
+			*ast.IntegerExpression,
+			*ast.FixedPointExpression,
+			*ast.PathExpression:
+			break
+		case *ast.IdentifierExpression:
+			// these are guaranteed to exist at time of destruction, so we allow them
+			if arg.Identifier.Identifier == SelfIdentifier || arg.Identifier.Identifier == BaseIdentifier {
+				break
+			}
+			// if it's an attachment, then it's also okay
+			if checker.typeActivations.Find(arg.Identifier.Identifier) != nil {
+				break
+			}
+			checker.report(&DefaultDestroyInvalidArgumentError{
+				Range: ast.NewRangeFromPositioned(checker.memoryGauge, arg),
+			})
+		case *ast.MemberExpression:
+			checkParamExpressionKind(arg.Expression)
+		case *ast.IndexExpression:
+			checkParamExpressionKind(arg.TargetExpression)
+			checkParamExpressionKind(arg.IndexingExpression)
+
+			// indexing expressions on arrays can fail, and must be disallowed, but
+			// indexing expressions on dicts, or composites (for attachments) will return `nil` and thus never fail
+			targetExprType := checker.Elaboration.IndexExpressionTypes(arg).IndexedType
+			switch targetExprType.(type) {
+			case ArrayType:
```

Comment:
Maybe flip the logic here and disallow by default, and allow only known cases, like mentioned in the comment